src/Dispatch.elm

```diff
@@ -0,0 +1,262 @@
+   and return the applied results as a list
+-}
+applyMultipleDecoders : List (Json.Decoder m) -> Json.Decoder (List m)
+applyMultipleDecoders decoders =
+  let
+    processDecoder initial decoder =
+      Json.decodeValue decoder initial
+        |> Result.toMaybe
+  in
+    Json.customDecoder Json.value
+      (\initial ->
+        List.map (processDecoder initial) decoders
+          |> List.filterMap identity
+          |> Result.Ok
+      )
+
+
+{-|
+-}
+forwardDecoder : List (Json.Decoder a) -> Json.Decoder (Msg a)
+forwardDecoder =
+  applyMultipleDecoders
+
+
+{-| Run multiple decoders on a single Html Event
+-}
+onEvt :
+  (Msg msg -> msg)
+  -> String
+  -> List (Json.Decoder msg)
+  -> (Html.Attribute msg)
+onEvt lift event =
+  onEvtOptions lift event Html.Events.defaultOptions
+
+
+{-| Run multiple decoders on a single Html Event with
+the given options
+-}
+onEvtOptions :
+  (Msg msg -> msg)
+  -> String
+  -> Html.Events.Options
+  -> List (Json.Decoder msg)
+  -> (Html.Attribute msg)
+onEvtOptions lift event options decoders =
+  case decoders of
+    [] ->
+      noop
+
+    [ x ] ->
+      Html.Events.onWithOptions event options x
+
+    _ ->
+      forwardDecoder decoders
+        |> Json.map lift
+        |> Html.Events.onWithOptions event options
+
+
+{-| A single event
+-}
+onSingle :
+  String
+  -> Html.Events.Options
+  -> List (Json.Decoder msg)
+  -> (Html.Attribute msg)
+onSingle event options decoders =
+  case decoders of
+    [] ->
+      noop
+
+    [ x ] ->
+      Html.Events.onWithOptions event options x
+
+    x :: xs ->
+      let
+        -- NOTE: This probably needs to be changed, currently only for debugging
+        _ = Debug.log "WARNING" ("Multiple decoders for Event '" ++ event ++ "' with no `Options.dispatch Mdl`")
+      in
+        Html.Events.onWithOptions event options x
+
+
+pickOptions : List (Decoder a) -> Html.Events.Options
+pickOptions decoders =
+  List.map snd decoders
+    |> List.filterMap identity
+    |> List.head
+    |> Maybe.withDefault Html.Events.defaultOptions
+
+
+{-| This function takes a lifting function and
+pair of events and their handlers and returns
+a list of `Html.Attribute` containing handlers that
+will allow for dispatching of multiple events from a single `Html.Event`
+-}
+listeners :
+  (Msg a -> a)
+  -> List ( String, List (Decoder a) )
+  -> List (Html.Attribute a)
+listeners lift items =
+  items
+    |> List.map (\( event, decoders ) ->
+       onEvtOptions lift event (pickOptions decoders) (List.map fst decoders)
+       )
+
+
+{-| Take a list of events to a list of decoders and
+apply only the first of those decoders. This
+is only applicable if no lifting argument is available.
+If a lifting argument is available use
+`Dispatch.listeners` instead.
+-}
+listeners' :
+  List ( String, List (Decoder a) )
+  -> List (Html.Attribute a)
+listeners' items =
+  items
+    |> List.map
+        (\( event, decoders ) ->
```

Comment:
Lose the closure—change the signature of `onSingle` to take `(event, decoders)` directly; do `pickOption` itself, and avoid the `List.map fst decoders` below by applying `fst` inside `onSingle`. 

Also consider trying specialising `onSingle` for the common case of one event-one decoder.
 

(I think this'll help.)